File: agents/producer/dashboard/duration_plan.py
"""Yuvo Studio — multi-clip ad duration planning model (Python mirror).

Pure module: no I/O, no network, no Supabase. Deterministic. Mirrors
web/lib/planning/duration-plan.ts so the runner side (Python) and the
dashboard side (TypeScript) agree byte-for-byte on the clip split.

Product decision: a 25-30s ad is PLANNED whole up front, then realised
as N connected clips. We never generate 15s and randomly extend.

    15s -> 1 clip  [15]
    20s -> 2 clips [15, 5]
    25s -> 2 clips [15, 10]
    30s -> 2 clips [15, 15]

Pai V4 lesson baked in: products with label-hallucination risk get a
blank-label / label-stripped reference and "product as soft prop, no
readable label text".

NO paid call is ever made by this module.
"""
from __future__ import annotations

from dataclasses import dataclass
from typing import Literal

TargetDuration = Literal[15, 20, 25, 30]
TARGET_DURATION_OPTIONS: tuple[int, ...] = (15, 20, 25, 30)

# Per-second credit estimate by tier (mirrors web/lib/quality-tiers.ts).
_PER_SECOND_CREDITS: dict[str, float] = {
    "draft_480p": 82.0,
    "standard_720p": 176.4,
    "premium_1080p": 396.0,
}

_CLIP_DURATIONS: dict[int, list[int]] = {
    15: [15],
    20: [15, 5],
    25: [15, 10],
    30: [15, 15],
}


@dataclass(frozen=True)
class PlannedClip:
    clip_number: int
    duration_seconds: int
    purpose: str
    continuation_role: str  # standalone | open_loop | close_loop
    script_window_start_sec: int
    script_window_end_sec: int
    script_segment: str
    visual_direction: str
    continuity_notes: str
    stitch_notes: str
    estimated_credits: int


@dataclass(frozen=True)
class AdDurationPlan:
    target_duration_seconds: int
    clip_strategy: str  # single_clip | multi_clip_stitched
    clip_count: int
    clips: list[PlannedClip]
    shared_negative_prompt: str
    blank_label_required: bool
    product_reference_rationale: str
    total_estimated_credits: int
    generation_count: int
    stitch_method: str = "concat"
    stitch_transition: str = "hard_cut_or_natural_pause"
    stitch_audio_strategy: str = "native_audio_per_clip_then_review"
    stitch_final_asset_kind: str = "stitched_video"


def _estimate_credits(tier_id: str, duration_sec: int) -> int:
    per_sec = _PER_SECOND_CREDITS.get(tier_id, _PER_SECOND_CREDITS["standard_720p"])
    return round(per_sec * duration_sec)


def _ad_beats(target: int) -> list[tuple[int, int, str]]:
    if target == 15:
        return [
            (0, 3, "relatable hook"),
            (3, 10, "creator context + soft proof"),
            (10, 15, "natural closing line"),
        ]
    return [
        (0, 5, "relatable hook / problem"),
        (5, 12, "creator explains context"),
        (12, 15, "soft transition (NOT an ending)"),
        (15, max(22, target - 8), "routine / application / proof continuation"),
        (max(22, target - 8), target - 3, "benefit / reason-to-believe"),
        (target - 3, target, "natural CTA / closing line"),
    ]


def _beats_in_window(
    beats: list[tuple[int, int, str]], start: int, end: int
) -> list[str]:
    return [label for (s, e, label) in beats if s < end and e > start]


def plan_ad_duration(
    *,
    target_duration_seconds: int,
    quality_tier: str = "standard_720p",
    shared_negative_prompt: str = "",
    label_hallucination_risk: bool = False,
) -> AdDurationPlan:
    """Build the deterministic clip plan for a target duration.

    Raises ValueError on an unsupported duration so callers fail loudly
    rather than silently producing a 15s ad.
    """
    if target_duration_seconds not in _CLIP_DURATIONS:
        raise ValueError(
            f"unsupported target_duration_seconds={target_duration_seconds!r}; "
            f"must be one of {TARGET_DURATION_OPTIONS}"
        )

    durations = _CLIP_DURATIONS[target_duration_seconds]
    clip_strategy = "single_clip" if len(durations) == 1 else "multi_clip_stitched"
    beats = _ad_beats(target_duration_seconds)

    clips: list[PlannedClip] = []
    cursor = 0
    for idx, dur in enumerate(durations):
        clip_number = idx + 1
        start_sec = cursor
        end_sec = cursor + dur
        cursor = end_sec

        is_only = len(durations) == 1
        is_last = clip_number == len(durations)
        role = (
            "standalone"
            if is_only
            else "close_loop"
            if is_last
            else "open_loop"
        )
        window_beats = _beats_in_window(beats, start_sec, end_sec)
        beat_summary = "; ".join(window_beats)

        purpose = (
            "Full self-contained 15s UGC ad"
            if is_only
            else "Hook + context - opens a loop, does NOT resolve the ad"
            if clip_number == 1
            else "Routine/proof continuation + natural CTA - closes the loop"
        )

        if role == "open_loop":
            tail = (
                "End on an unfinished thought / soft transition - clip 2 "
                "must feel like the same take continuing, not a new ad."
            )
        elif role == "close_loop":
            tail = (
                "Pick up exactly where clip 1 left off (same sentence "
                "energy, same setting/lighting/creator), then land the CTA."
            )
        else:
            tail = "Self-contained: hook -> context -> soft close."
        script_segment = (
            f"[Full-ad timeline {start_sec}-{end_sec}s] {beat_summary}. {tail}"
        )

        visual_direction = (
            "Same creator, room, lighting, wardrobe as every other clip. "
            "Real-friend UGC register, handheld feel. Product is a soft "
            "prop, never a label-hero shot. "
            + (
                "Establish the creator + setting in the first 2s."
                if clip_number == 1
                else "Match clip 1's framing + colour exactly so the cut "
                "is invisible."
            )
        )

        if is_only:
            continuity = "Single clip - no cross-clip continuity needed."
        elif clip_number == 1:
            continuity = (
                "Lock creator identity, wardrobe, room, lighting, focal "
                "length, and product-reference image - clip 2 reuses ALL."
            )
        else:
            continuity = (
                "MUST reuse clip 1's exact creator/wardrobe/room/lighting "
                "and the SAME product reference image. No new location, "
                "outfit, or time-of-day change."
            )

        if is_only:
            stitch = "No stitch - single deliverable."
        elif is_last:
            stitch = (
                "Concatenate after clip 1. Expect a hard cut or natural "
                "pause at the seam; review audio before client share."
            )
        else:
            stitch = (
                "This clip ends mid-thought; the seam to clip 2 should "
                "land on a natural pause/breath, not a finished sentence."
            )

        clips.append(
            PlannedClip(
                clip_number=clip_number,
                duration_seconds=dur,
                purpose=purpose,
                continuation_role=role,
                script_window_start_sec=start_sec,
                script_window_end_sec=end_sec,
                script_segment=script_segment,
                visual_direction=visual_direction,
                continuity_notes=continuity,
                stitch_notes=stitch,
                estimated_credits=_estimate_credits(quality_tier, dur),
            )
        )

    total = sum(c.estimated_credits for c in clips)
    rationale = (
        "Pai V4 lesson: prompt-only label guardrails FAILED; a "
        "blank-label / label-stripped product reference SUCCEEDED. Use a "
        "label-stripped reference image, product as soft prop, verify no "
        "readable label text before client share."
        if label_hallucination_risk
        else "No known label-hallucination risk. Standard reference "
        "acceptable; still keep product a soft prop, avoid packaging-text "
        "close-ups."
    )

    return AdDurationPlan(
        target_duration_seconds=target_duration_seconds,
        clip_strategy=clip_strategy,
        clip_count=len(clips),
        clips=clips,
        shared_negative_prompt=shared_negative_prompt,
        blank_label_required=label_hallucination_risk,
        product_reference_rationale=rationale,
        total_estimated_credits=total,
        generation_count=len(clips),
    )


def to_clip_plan_json(plan: AdDurationPlan) -> dict:
    """Compact JSON for generation_batches.clip_plan (migration 007)."""
    return {
        "target_duration_seconds": plan.target_duration_seconds,
        "clip_strategy": plan.clip_strategy,
        "clips": [
            {
                "clip_number": c.clip_number,
                "duration_seconds": c.duration_seconds,
                "purpose": c.purpose,
                "continuation_role": c.continuation_role,
            }
            for c in plan.clips
        ],
        "stitch_plan": {
            "method": plan.stitch_method,
            "transition": plan.stitch_transition,
            "audio_strategy": plan.stitch_audio_strategy,
            "final_asset_kind": plan.stitch_final_asset_kind,
        },
        "product_reference_strategy": {
            "blank_label_required": plan.blank_label_required,
        },
        "estimated_credits_total": plan.total_estimated_credits,
        "generation_count": plan.generation_count,
    }

File: web/lib/planning/duration-plan.ts
// Yuvo Studio — multi-clip ad duration planning model.
//
// PURE module. No React, no I/O, no Supabase, no env. Importable from
// both the operator UI (preview) and any future server action (persist
// the plan). Deterministic: same input → same plan.
//
// Product decision (do NOT change without operator sign-off):
//   We never "generate 15s then randomly extend". A 25–30s ad is
//   PLANNED as a whole up front, then realised as multiple connected
//   clips. Enhancor/Seedance generates at most ~15s of clean UGC per
//   call, so longer ads are N connected clips stitched after the fact.
//
//   15s  → 1 clip  [15]
//   20s  → 2 clips  [15, 5]
//   25s  → 2 clips  [15, 10]
//   30s  → 2 clips  [15, 15]
//
// The Pai V4 lesson is baked in: for products with label-hallucination
// risk the plan's product_reference_strategy defaults to a
// blank-label / label-stripped reference, NOT a sharp packshot, and the
// product is treated as a soft prop with no readable label text.

import { estimateGenerationCredits } from "@/lib/quality-tiers";
import type { QualityTierId } from "@/lib/types";

export type TargetDurationSeconds = 15 | 20 | 25 | 30;

export const TARGET_DURATION_OPTIONS: readonly TargetDurationSeconds[] = [
  15, 20, 25, 30,
] as const;

export type ClipStrategy = "single_clip" | "multi_clip_stitched";

/** A clip's narrative job inside the whole ad. `open_loop` deliberately
 *  does NOT resolve the ad (clip 2 follows); `close_loop` lands the CTA;
 *  `standalone` is the 15s single-clip case. */
export type ContinuationRole = "standalone" | "open_loop" | "close_loop";

export interface PlannedClip {
  clipNumber: number;
  durationSeconds: number;
  /** Short human label of what this clip does in the ad. */
  purpose: string;
  continuationRole: ContinuationRole;
  /** The slice of the overall script this clip covers, expressed as a
   *  time window over the FULL ad timeline (not the clip's local time). */
  scriptWindow: { startSec: number; endSec: number };
  /** Operator-editable scaffold; the real copy is authored in the
   *  prompt editor. These are starting points, not final prompts. */
  scriptSegment: string;
  visualDirection: string;
  continuityNotes: string;
  stitchNotes: string;
  /** Per-clip credit estimate at the chosen tier. */
  estimatedCredits: number;
}

export interface StitchPlan {
  method: "concat";
  transition: "hard_cut_or_natural_pause";
  audioStrategy: "native_audio_per_clip_then_review";
  /** Final asset kind once all clips are stitched + operator-approved. */
  finalAssetKind: "stitched_video";
  notes: string;
}

export interface ProductReferenceStrategy {
  /** When true, the plan instructs use of a blank-label / label-stripped
   *  reference image and "product as soft prop, no readable label". */
  blankLabelRequired: boolean;
  rationale: string;
}

export interface AdDurationPlan {
  targetDurationSeconds: TargetDurationSeconds;
  clipStrategy: ClipStrategy;
  clipCount: number;
  clips: PlannedClip[];
  /** Shared across every clip — same creator, setting, tone, product
   *  rules. Authored once in the prompt editor, applied to all clips. */
  sharedNegativePrompt: string;
  productReferenceStrategy: ProductReferenceStrategy;
  stitchPlan: StitchPlan;
  /** Sum of every clip's estimate. The operator pays per generation —
   *  this is N generations, never "one 30s API call". */
  totalEstimatedCredits: number;
  /** How many paid Seedance generations this plan implies. */
  generationCount: number;
}

// ---------------------------------------------------------------------------
// Clip duration table — the single source of truth for the split.
// ---------------------------------------------------------------------------
const CLIP_DURATIONS: Record<TargetDurationSeconds, number[]> = {
  15: [15],
  20: [15, 5],
  25: [15, 10],
  30: [15, 15],
};

/** Narrative skeleton for the WHOLE ad, expressed over the full
 *  timeline. The clip planner slices this by each clip's window so
 *  clip 2 reads as a continuation, not a fresh ad. Mirrors the 30s
 *  structure in docs/multi_clip_ad_duration_plan.md. */
interface Beat {
  startSec: number;
  endSec: number;
  label: string;
}

function adBeats(target: TargetDurationSeconds): Beat[] {
  if (target === 15) {
    return [
      { startSec: 0, endSec: 3, label: "relatable hook" },
      { startSec: 3, endSec: 10, label: "creator context + soft proof" },
      { startSec: 10, endSec: 15, label: "natural closing line" },
    ];
  }
  // 20 / 25 / 30 — open-loop clip 1, continuation clip 2.
  return [
    { startSec: 0, endSec: 5, label: "relatable hook / problem" },
    { startSec: 5, endSec: 12, label: "creator explains context" },
    { startSec: 12, endSec: 15, label: "soft transition (NOT an ending)" },
    {
      startSec: 15,
      endSec: Math.max(22, target - 8),
      label: "routine / application / proof continuation",
    },
    {
      startSec: Math.max(22, target - 8),
      endSec: target - 3,
      label: "benefit / reason-to-believe",
    },
    { startSec: target - 3, endSec: target, label: "natural CTA / closing line" },
  ];
}

function beatsInWindow(beats: Beat[], startSec: number, endSec: number): Beat[] {
  return beats.filter((b) => b.startSec < endSec && b.endSec > startSec);
}

// ---------------------------------------------------------------------------
// Public API
// ---------------------------------------------------------------------------

export interface PlanInput {
  targetDurationSeconds: TargetDurationSeconds;
  qualityTier: QualityTierId;
  /** Carried from the prompt version. The planner does NOT invent ad
   *  copy — it scaffolds windows; the operator authors the real text. */
  sharedNegativePrompt: string;
  /** Pai V4 learning: set true for products whose real packaging has
   *  readable copy the model will hallucinate (skincare with dense
   *  labels, etc.). Drives the blank-label reference instruction. */
  labelHallucinationRisk: boolean;
}

export function planAdDuration(input: PlanInput): AdDurationPlan {
  const { targetDurationSeconds, qualityTier } = input;
  const durations = CLIP_DURATIONS[targetDurationSeconds];
  const clipStrategy: ClipStrategy =
    durations.length === 1 ? "single_clip" : "multi_clip_stitched";
  const beats = adBeats(targetDurationSeconds);

  let cursor = 0;
  const clips: PlannedClip[] = durations.map((dur, idx) => {
    const clipNumber = idx + 1;
    const startSec = cursor;
    const endSec = cursor + dur;
    cursor = endSec;

    const isOnlyClip = durations.length === 1;
    const isLastClip = clipNumber === durations.length;
    const continuationRole: ContinuationRole = isOnlyClip
      ? "standalone"
      : isLastClip
        ? "close_loop"
        : "open_loop";

    const windowBeats = beatsInWindow(beats, startSec, endSec);
    const beatSummary = windowBeats.map((b) => b.label).join("; ");

    const purpose = isOnlyClip
      ? "Full self-contained 15s UGC ad"
      : clipNumber === 1
        ? "Hook + context — opens a loop, does NOT resolve the ad"
        : "Routine/proof continuation + natural CTA — closes the loop";

    const scriptSegment =
      `[Full-ad timeline ${startSec}-${endSec}s] ${beatSummary}. ` +
      (continuationRole === "open_loop"
        ? "End on an unfinished thought / soft transition — clip 2 must " +
          "feel like the same take continuing, not a new ad."
        : continuationRole === "close_loop"
          ? "Pick up exactly where clip 1 left off (same sentence energy, " +
            "same setting/lighting/creator), then land the CTA naturally."
          : "Self-contained: hook → context → soft close.");

    const visualDirection =
      "Same creator, same room, same lighting, same wardrobe as every " +
      "other clip. Real-friend UGC register, handheld feel. Product is a " +
      "soft prop, never a label-hero shot. " +
      (clipNumber === 1
        ? "Establish the creator + setting in the first 2s."
        : "Match clip 1's framing + colour exactly so the cut is invisible.");

    const continuityNotes =
      isOnlyClip
        ? "Single clip — no cross-clip continuity needed."
        : clipNumber === 1
          ? "Lock creator identity, wardrobe, room, lighting, focal length, " +
            "and product-reference image — clip 2 reuses ALL of these."
          : "MUST reuse clip 1's exact creator/wardrobe/room/lighting and " +
            "the SAME product reference image. No new location, no new " +
            "outfit, no time-of-day change.";

    const stitchNotes = isOnlyClip
      ? "No stitch — single deliverable."
      : isLastClip
        ? "Concatenate after clip 1. Expect a hard cut or a natural " +
          "pause at the seam; review audio continuity before client share."
        : "This clip ends mid-thought; the seam to clip 2 should land on " +
          "a natural pause/breath, not a finished sentence.";

    return {
      clipNumber,
      durationSeconds: dur,
      purpose,
      continuationRole,
      scriptWindow: { startSec, endSec },
      scriptSegment,
      visualDirection,
      continuityNotes,
      stitchNotes,
      estimatedCredits: estimateGenerationCredits(qualityTier, dur),
    };
  });

  const totalEstimatedCredits = clips.reduce(
    (sum, c) => sum + c.estimatedCredits,
    0,
  );

  const productReferenceStrategy: ProductReferenceStrategy = {
    blankLabelRequired: input.labelHallucinationRisk,
    rationale: input.labelHallucinationRisk
      ? "Pai V4 lesson: prompt-only label guardrails FAILED; a " +
        "blank-label / label-stripped product reference SUCCEEDED. Use a " +
        "label-stripped reference image, treat the product as a soft " +
        "prop, and verify no readable label text before client share."
      : "No known label-hallucination risk for this product. Standard " +
        "reference acceptable, but still keep the product a soft prop " +
        "and avoid packaging-text close-ups.",
  };

  const stitchPlan: StitchPlan = {
    method: "concat",
    transition: "hard_cut_or_natural_pause",
    audioStrategy: "native_audio_per_clip_then_review",
    finalAssetKind: "stitched_video",
    notes:
      clipStrategy === "single_clip"
        ? "Single clip — the raw_video IS the deliverable; no stitch step."
        : "When ALL clips reach status=completed, concat in clip order " +
          "into one final asset (kind=stitched_video). Raw clips stay as " +
          "raw_video assets. client_safe_video_url only points at the " +
          "stitched final AFTER operator approval + label-readability check.",
  };

  return {
    targetDurationSeconds,
    clipStrategy,
    clipCount: clips.length,
    clips,
    sharedNegativePrompt: input.sharedNegativePrompt,
    productReferenceStrategy,
    stitchPlan,
    totalEstimatedCredits,
    generationCount: clips.length,
  };
}

/** Serialise to the JSON shape stored in
 *  `generation_batches.clip_plan` (migration 007). Kept separate from
 *  the rich planning object so the DB column stays compact + stable. */
export function toClipPlanJson(plan: AdDurationPlan): Record<string, unknown> {
  return {
    target_duration_seconds: plan.targetDurationSeconds,
    clip_strategy: plan.clipStrategy,
    clips: plan.clips.map((c) => ({
      clip_number: c.clipNumber,
      duration_seconds: c.durationSeconds,
      purpose: c.purpose,
      continuation_role: c.continuationRole,
    })),
    stitch_plan: {
      method: plan.stitchPlan.method,
      transition: plan.stitchPlan.transition,
      audio_strategy: plan.stitchPlan.audioStrategy,
      final_asset_kind: plan.stitchPlan.finalAssetKind,
    },
    product_reference_strategy: {
      blank_label_required: plan.productReferenceStrategy.blankLabelRequired,
    },
    estimated_credits_total: plan.totalEstimatedCredits,
    generation_count: plan.generationCount,
  };
}
